### PR TITLE
Feature/distance calculator

### LINKED ISSUE #3: Implement a distance calculator
Implement a DistanceMatrix class that takes a sparse graph representing the real public transport network, and has methods to calculate distances between every nodes, effectively completing the graph. It also has to store intermediate paths for later use (ex : store that the shortest path to do A->D is A->B->C->D). To do that, use algorithms like Dijkstra or BFS.

### LINKED ISSUE #10: Add tests for the distance calculator
Add tests for the distance calculator : tests on simple graphs that can be visualised + edge cases.

### PR DESCRIPTION
**Pull request by Polemique**
Distance calculator based on strategies

### CODE REVIEW

File: src/hamilbus/__init__.py [hunk 1]
```diff
@@ -2,15 +2,18 @@
 from .datamodels import BusNetworkGraph, Line, Stop
 from .reader import load_lines, load_stops, parse_stop_id
 from .web import app, run_server
+from .distance_matrix import compute_distance_matrix
 
 __all__ = [
     "BusNetworkGraph",
     "Line",
     "Stop",
+    "BusNetworkGraph",
```
Comment (KarlLbs) [inline src/hamilbus/__init__.py]: It's already there 3 lines above

File: src/hamilbus/distance_matrix.py [hunk 2]
```diff
@@ -1,2 +1,28 @@
...
+    bus_graph: BusNetworkGraph, strategy: str = "dijkstra"
+) -> dict:
+    """Computes the shortest distance and path between all pairs of stops in the graph."""
+    distance_matrix = {}
+    strategy_func = {
+        "dijkstra": nx.single_source_dijkstra,
+        "bellman-ford": nx.single_source_bellman_ford,
+    }
+
+    for u in bus_graph.graph.nodes:
```
Comment (KarlLbs) [inline src/hamilbus/distance_matrix.py]: The "bus_graph.graph" makes me wonder : instead of having a dataclass with only attribute the nx graph, wouldn't it be cleaner to have a class inherited from the nx.Multigraph directly, to which we override certain methods and add others ? This would allow us to avoid things like bus_graph.graph.something and allow us to directly use built-in functions like has_edge(). 
Just a suggestion, tell me what you think @Polemique

**Review by KarlLbs — APPROVED** [review]
All good 👍

File: src/hamilbus/distance_matrix.py [hunk 3]
```diff
@@ -1,2 +1,28 @@
...
+        for v in bus_graph.graph.nodes:
+            if u != v:
+                try:
+                    length, path = strategy_func[strategy](
+                        bus_graph.graph, u, v, weight="distance"
+                    )
+                    distance_matrix[(u, v)] = {"distance": length, "path": path}
```
Comment (KarlLbs) [inline src/hamilbus/distance_matrix.py]: I think most solvers that we will use after this step will require a distance matrix in the format list(list())/array of n lines and m columns, where for each line n, the m-th value is the distance from node n to node m. For example, let's say we have nodes A, B, C, the distance matrix will be :
[ [distance(A, A),   distance(A, B),   distance(A, C)],
  [distance(B, A),   distance(B, B),   distance(B, C)],
  [distance(C, A),   distance(C, B),   distance(C, C)] ]
With the diagonal being 0s.
So it might be a good idea to build that directly, what do you think @Polemique ?
Maybe we can build one array containing the distances and another one, similar, containing the paths ? 
That will avoid having to do a transformation on this {"distance": length, "path": path} structure later in the code.

**Review by KarlLbs — CHANGES_REQUESTED** [review]

File: src/hamilbus/distance_matrix.py (hunk 3 above)
Comment (KarlLbs) [reply src/hamilbus/distance_matrix.py]: Another thing to note : once we have that array I described, we'll need to now which row/column of the array corresponds to which stop id, because the algorithms will give us a path referencing the rows/columns numbers of the matrix. So if you could build a dictionnary {index in the matrix : stop id, ...} during the for loops, that would be great ! @Polemique

File: src/hamilbus/distance_matrix.py (hunk 2 above)
Comment (Polemique) [reply src/hamilbus/distance_matrix.py]: Could be a refactor yes ! I like the idea, just could we keep in the refactor the naming ith stops/lines in methods names ? 

Can you create the issue ?

**Comment by Polemique** [comment]
@KarlLbs second pass ? :)

**Review by KarlLbs — APPROVED** [review]
All good 👍
